pkg/auth/auth.go

```diff
@@ -51,51 +57,102 @@ func decode64(s string) (string, error) {
+}
+
+// ReadUserGrpcOrDie should only be used in the cd-service.
+// ReadUserGrpcOrDie takes the User from middleware (context).
+// It returns a User or an error if the user is not found.
+func ReadUserGrpcOrDie(ctx context.Context) (*User, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, httperrors.AuthError(ctx, errors.New("could not retrieve metadata context with git author in grpc context"))
+	}
+	originalEmailArr := md.Get("author-email")
+	if len(originalEmailArr) == 0 {
+		return nil, httperrors.AuthError(ctx, errors.New("did not find author-email in grpc context"))
+	}
+	originalEmail := originalEmailArr[0]
+	userMail, err := decode64(originalEmail)
+	if err != nil {
+		return nil, httperrors.AuthError(ctx, fmt.Errorf("extract: non-base64 in author-email in grpc context %s", originalEmail))
+	}
+	originalNameArr := md.Get("author-username")
+	if len(originalNameArr) == 0 {
+		return nil, httperrors.AuthError(ctx, fmt.Errorf("extract: username undefined but mail defined in grpc context %s", userMail))
+	}
+	originalName := originalNameArr[0]
+	userName, err := decode64(originalName)
+	if err != nil {
+		return nil, httperrors.AuthError(ctx, fmt.Errorf("extract: non-base64 in author-username in grpc context %s", userName))
+	}
+	logger.FromContext(ctx).Warn(fmt.Sprintf("Extract: original mail %s. Decoded: %s", originalEmail, userMail))
+	logger.FromContext(ctx).Warn(fmt.Sprintf("Extract: original name %s. Decoded: %s", originalName, userName))
+	u := &User{
+		Email: userMail,
+		Name:  userName,
+	}
+	if u.Email == "" || u.Name == "" {
+		return nil, httperrors.AuthError(ctx, errors.New("email and name in grpc context cannot both be empty"))
+	}
+	return u, nil
+}
+
+// ReadUserHttpOrDie should only be used in the cd-service.
+// ReadUserHttpOrDie takes the User from the http request.
+// It returns a User or an error if the user is not found.
+func ReadUserHttpOrDie(ctx context.Context, r *http.Request) (*User, error) {
+	headerEmail, err := decode64(r.Header.Get(HttpHeaderUserEmail))
+	if err != nil {
+		return nil, httperrors.AuthError(ctx, errors.New("ExtractUserHttp: invalid data in email"))
+	}
+	headerName, err := decode64(r.Header.Get(HttpHeaderUserName))
+	if err != nil {
+		return nil, httperrors.AuthError(ctx, errors.New("ExtractUserHttp: invalid data in name"))
+	}
+
+	if headerName != "" && headerEmail != "" {
+		return &User{
+			Email: headerEmail,
+			Name:  headerName,
+		}, nil
+	}
+	return nil, httperrors.AuthError(ctx, errors.New("ExtractUserHttp: did not find data in headers"))
+}
+
+// WriteUserToHttpHeader should only be used in the frontend-service
+// WriteUserToHttpHeader writes the user into http headers
+// it is used for requests like /release which are delegated from frontend-service to cd-service
+func WriteUserToHttpHeader(r *http.Request, user User, ctxLOG context.Context) {
+	r.Header.Set(HttpHeaderUserName, encode64(user.Name))
```

Comment:
Is there any deep reason why you don't use the same header here as for grpc?

Reply:
no. I'll change it to use the same

Reply:
done